### PR TITLE
Add reparent methods to Node

### LINKED ISSUE #153: Add reparenting method
**Describe the project you are working on:**
A Diablo-style 2D action RPG.

**Describe the problem or limitation you are having in your project:**
Any time I pass around nodes (for example, I have status effects which are transferred from an attack object to it's victim) reparenting them requires that I first get the existing parent, remove the child, and then add them to their new parent. This causes what feels like unnecessary code and scene tree traversal:
```
node.get_parent().remove_child(node)
new_parent.add_child(node)
```

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
A reparenting function built in to the engine would be faster than my GDScript and would save me from repeating most of the code in the example above every time I need to reparent.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
Implement a new `node.set_parent(new_parent)` method or change the existing `add_child(node)` method to remove any existing parents by default

**Is there a reason why this should be core and not an add-on in the asset library?:**
It is a common use case that would be faster if implemented in the engine.

### LINKED ISSUE #21781: Method to reparent Node while keeping global transform
Right now, if you want to change parent of a node, you need to do:
```
node.get_parent().remove_child(node)
new_parent.add_child(node)
```
This already causes problems, as I see people confused on why they can't just `add_child()` to another node (the fact that they don't read/can't understand the error they are show is another thing). It's even worse if you want to reparent a Node2D. When doing above code, the result usually is "why my node has disappeared??!!!" (this is people not understanding local transform, but whatever).

Anyways, with 2D and 3D nodes you need:
```
var old_position = node.global_position
node.get_parent().remove_child(node)
new_parent.add_child(node)
node.global_position = old_position
```
That's... too much writing for a simple thing. So I'm proposing this:
`Node.reparent(new_parent, keep_global_transform = true)`

What it does is changing the parent of a node to new one. In some cases, it allows to avoid temporary variables if you want to e.g. move to granpdarent. The method would be overloaded for CanvasItem and Spatial to automatically (and optionally) retain the global_transform, so the node stays in place, like it is in editor actually.

This is somewhat related to #8788, but I'm proposing a more universal and concrete thing.

### LINKED ISSUE #24433: parent property on Node
Much like how the `get_name()` and `get_owner()` method of Node was deprecated in favor of the `name`  and `owner` property, why not do the same for a `parent` property (or in addition to `get_parent()`)?

### PR DESCRIPTION
**Pull request by KoBeWi**
Closes https://github.com/godotengine/godot-proposals/issues/153
Closes #21781

### CODE REVIEW

**Comment by akien-mga** [comment]
We discussed this on IRC with @reduz who is opposed to the idea of exposing `get_parent()` as a property, since it's not a property of the Node itself but an information on relationships between nodes. Abstracting `add_child` (which has strong implications on the scene tree and the node's behavior) is not desired, and making it a getter-only property is also not considered worth it due to the fact that it's not a read-only property internally.

IRC discussion:
```
[14:55:59] <Akien> reduz: WDYT of #36301 ? We add discussion already and several PRs proposing stuff like reparent_child() method that properly handles the global transform for CanvasItem/Spatial, but we never reached a consensus.
[14:56:00] <IssueBot> #36301: Add parent property to Node class | https://git.io/Jv8xw
[14:56:28] <Akien> reduz: a plain `parent` property with documentation confirming that it will *not* preserve transform seems OK to me.
[14:59:19] <Akien> I'm not sure how useful it will be unless you reparent plain Nodes, but I guess it's nice to be able to use `.parent` as property.
[15:01:53] <reduz> Akien: I think they are different problems, so they need different solutions
[15:02:40] <reduz> Parent property does not make much sense to me and seems a lot less readable
[15:03:47] <reduz> If you want to reparent and keep transform, which seems its the actual problem, and something difficult to do for many users, we can add a method for this in the relevant classes
[15:03:50] <Akien> reduz: well I think the intent is mostly to use it as getter (if node == parent)
[15:04:14] <Akien> Like we have `owner` for example.
[15:04:27] <reduz> Like reparent_with_transform
[15:05:28] <reduz> Akien: yeah but there are not many getter only properties in godot, and setting the parent instead of add child is very unintuitive
[15:06:14] <reduz> "add_chid" is better because it is more obvious that the order matters
[15:06:56] <reduz> So I would only solve the actual problem
[15:07:23] <Akien> Well the problem this one tries to solve is #24433
[15:07:23] <IssueBot> #24433: parent property on Node | https://git.io/Jv8pT
[15:07:33] <Akien> We could make it a getter-only property like `root` for consistency.
[15:07:57] <reduz> Its not the same
[15:08:10] <reduz> So it does not need to be consistent
[15:08:53] <reduz> Name and owner only have effect on the node ittself, while parent does not
[15:09:01] <Akien> Well we did enforce a new property-based workflow in 3.0, hiding getters from the docs and completion, so it's normal that users want consistency here.
[15:09:08] <reduz> And making it getter only is confusing at best
[15:13:52] <reduz> Akien: it is perfectly consistent as is. Children are contained in the parent. So parent is not a property but just information
[15:14:30] <reduz> If it was a propertty, it would be inconsistent and easy to misunderstand
[15:16:22] <reduz> Properties are for datq containing within thr object that you can modify in the context of the object
[15:16:44] <reduz> Parent ia not the case
```

**Comment by akien-mga** [comment]
We discussed further with @reduz, and we agreed that adding a `Node::reparent(Node *new_parent)` method would be fine, so similar to `set_parent()` here, but it should raise an error if there is no pre-existing parent. I.e. `add_child()` should stay the main way to add a node to the scene tree - it can then be reparented with `reparent()` for convenience.

**Comment by akien-mga** [comment]
Sorry it's a fast moving discussion on IRC :P
```
16:01 <reduz> No. i think reparent makes more sense than making a parent property because node needs to have been added
16:01 <Akien> So it should ERR_FAIL if there's no parent?
16:01 <reduz> Yes
16:01 <Akien> Ok, makes sense.
16:05 <reduz> well we can make it not fail, users won't really use it that way anyway and if they want to there is no reason to forbid it
```

So basically this PR is more or less good, but we'd suggest renaming the method to `reparent` and not expose it as a property. It thus fixes godotengine/godot-proposals#153, but not #24433.

**Comment by KoBeWi** [comment]
Would adding `reparent_with_transform` here be fine?

**Comment by KoBeWi** [comment]
Ok, just added them as second commit. However Control doesn't have `set_global_transform` method, so I'm not sure how to keep its rotation and scale.

**Review by akien-mga — COMMENTED** [review]
After some discussion we reached a "well why not though we're not sure how useful that really is" consensus to merge this :)

But we also changed our mind again on the API and suggest actually implementing a virtual `Node::reparent(Node *p_parent, bool p_keep_global_transform = false)`, which you can override in `Node2D`, `Control` and `Node3D` to actually keep the transform.

**Comment by KoBeWi** [comment]
Done. `p_keep_global_transform` is true by default, because it's the reason why this method was created in the first place.

File: scene/main/node.cpp [hunk 1]
```diff
@@ -1435,6 +1435,13 @@ TypedArray<Node> Node::find_children(const String &p_pattern, const String &p_ty
 	return ret;
 }
 
+void Node::reparent(Node *p_parent, bool p_keep_global_transform) {
+	ERR_FAIL_NULL(p_parent);
+	ERR_FAIL_NULL_MSG(data.parent, "Node needs a parent to be reparented.");
```
Comment (kleonc) [inline scene/main/node.cpp:1440]: Another condition:
```
	ERR_FAIL_COND_MSG(p_parent == data.parent, "Can't reparent Node to its current parent.");
```

File: doc/classes/Node.xml [hunk 2]
```diff
@@ -604,6 +604,15 @@
 				Removes a node from a group. See notes in the description, and the group methods in [SceneTree].
 			</description>
 		</method>
+		<method name="reparent">
+			<return type="void" />
+			<argument index="0" name="new_parent" type="Node" />
+			<argument index="1" name="keep_global_transform" type="bool" default="true" />
+			<description>
+				Changes the parent of this [Node] to the [code]new_parent[/code]. The node needs to already have a parent.
+				If [code]keep_global_transform[/code] is [code]true[/code], the node's global transform will be preserved if supported. [Node2D], [Node3D] and [Control] support this argument (but [Control] keeps only position).
```
Comment (kleonc) [inline doc/classes/Node.xml:613]:
```suggestion
			<param index="0" name="new_parent" type="Node" />
			<param index="1" name="keep_global_transform" type="bool" default="true" />
			<description>
				Changes the parent of this [Node] to the [param new_parent]. The node needs to already have a parent.
				If [param keep_global_transform] is [code]true[/code], the node's global transform will be preserved if supported. [Node2D], [Node3D] and [Control] support this argument (but [Control] keeps only position).
```

File: scene/main/node.cpp (hunk 1 above)
Comment (KoBeWi) [reply scene/main/node.cpp:1440]: Doing this is technically correct, but ok.

Comment (kleonc) [reply scene/main/node.cpp:1440]: Hmm, maybe it should a no-op instead of error? :thinking:

Comment (KoBeWi) [reply scene/main/node.cpp:1440]: Makes sense.

File: scene/main/node.cpp [hunk 3]
```diff
@@ -1435,6 +1435,18 @@ TypedArray<Node> Node::find_children(const String &p_pattern, const String &p_ty
...
```
Comment (Maran23) [inline scene/main/node.cpp:1440]: I don't have a strong opinion here but do we need this check?
I know that the `reparent` method does not really make sense, if the `node` does not have a parent (one can just use `add_child`.
But it is not harmful either.

File: scene/gui/control.cpp [hunk 4]
```diff
@@ -184,6 +184,14 @@ Size2 Control::_edit_get_minimum_size() const {
 }
 #endif
 
+void Control::reparent(Node *p_parent, bool p_keep_global_transform) {
+	Transform2D temp = get_global_transform();
+	Node::reparent(p_parent);
+	if (p_keep_global_transform) {
+		set_global_position(temp.get_origin());
```
Comment (Maran23) [inline scene/gui/control.cpp:191]: Should we also set `rotation` and `scale`?

**Review by Maran23 — COMMENTED** [review]
Got two questions, but otherwise everything works as expected.

File: scene/gui/control.cpp (hunk 4 above)
Comment (KoBeWi) [reply scene/gui/control.cpp:191]: So, the funny thing is that Control has no `set_global_transform()` method and no methods for global rotation and global scale. Hence only position can be set easily.

File: scene/main/node.cpp (hunk 3 above)
Comment (KoBeWi) [reply scene/main/node.cpp:1440]: Well, it's not harmful, but the method is meant for reparenting. As you said, one can just use `add_child()`.

**Review by akien-mga — APPROVED** [review]

**Comment by Flavelius** [comment]
Whenever i see arguments around these workarounds i wonder why there's not just a set_parent(bool:keep_transform) method (like in unity for example). 
add_child fails anyway if the added node currently already has a parent, in which case it's the safety-preferred way to always check for an existing parent first via get_parent() (and handle that case). 
This makes multiple operations replace what could be a single, simple set_parent(keep_transform) which is expressive enough in my opinion to indicate that it handles all parenting concerns (including existing parents). 
And as the (existing) parent handling is then done on the engine side (currently anyway for error checks) it's likely even a minor performance gain in not having to duplicate the redundant check (and the potential unparenting too) on the scripting side.
reparent(..) seems just like an additional unnecessary in-between step, because it fails on missing existing parents (similar to the opposite on add_child), so that check still has to be duplicated to the scripting side and be handled according, potentially leading to even more code inside conditionals.

**Comment by YuriSizov** [comment]
@Flavelius Where in your proposal do you actually specify which node should be a parent?

**Comment by Flavelius** [comment]
> @Flavelius Where in your proposal do you actually specify which node should be a new parent?

Yes, sorry, this not a complete proposal, just something I wanted to note for consideration. The full method signature could look like this set_parent(parent_node, keep_transform), similar to the one unity has (https://docs.unity3d.com/ScriptReference/Transform.SetParent.html)

**Comment by YuriSizov** [comment]
@Flavelius It's the same signature as the new `Node::reparent` has too. So what's the idea behind your proposal then? I don't quite understand it.

**Comment by KoBeWi** [comment]
I think the idea is to unify `add_child()` and `reparent()` into a single `set_parent()` method. Right now both are very similar, but `add_child()` fails if node has a parent and `reparent()` fails if node has no parent.

**Comment by KevinStirling** [comment]
As of godot 4 alpha 14 the reparent() method appears to now be missing, possibly a regression?

**Comment by kleonc** [comment]
> As of godot 4 alpha 14 the reparent() method appears to now be missing, possibly a regression?

No, `4.0.alpha14` is over five months old, `4.0.beta13` is the newest version, [check them out in here](https://downloads.tuxfamily.org/godotengine/4.0/).

**Comment by KevinStirling** [comment]
> 

Oh gosh, i clicked the wrong link from google when i went to get the new beta and didn't even realize i was using the alpha. thank you!

**Pull request merged**